rep-2004.rst

```diff
@@ -0,0 +1,527 @@
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+*Note: bullets below that start with [ROS Core], will be the prescription for what we do in the core packages in order to meet the associated requirements*
+
+Requirements to be considered a 'Level 4' package:
+
+* Version Policy:
+
+  * No requirements, but having a policy is still recommended (e.g. ``semver``), even if the version is not yet stable (e.g. >= 1.0.0 for ``semver``)
+
+* Change Control Process:
+
+  * No explicit change control process required, but still recommended
+
+* Documentation:
+
+  * Must have a declared license or set of licenses
+  * Must have a copyright statement in each source file
+  * [ROS Core]:
+
+    * Must have automated checks for copyright statements and licenses
+    * Must use the Apache 2.0 license, unless the package has an existing permissive license (e.g. rviz uses three-clause BSD)
+
+* Testing:
+
+  * No explicit testing requirements, though covering some if not all of the features with tests is recommended
+
+* Dependencies:
+
+  * No restrictions
+
+* Platform Support:
+
+  * May support all tier 1 platforms for ROS 2, as defined in `REP-2000 <https://www.ros.org/reps/rep-2000.html#support-tiers>`_
+
+Any package that does not claim to be 'Level 3' or higher is automatically 'Level 4'.
+Refer to the detailed description of the requirements in the Quality Level 1 section above for more information.
+
+Quality Level 5
+^^^^^^^^^^^^^^^
+
+Packages in this category simply do not meet even the 'Level 4' requirements, and for that reason should not be used.
+The rationale being that all packages should have at least a declare license or licenses and should include copyright statements in each file.
+
+Repository Organization
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Since these categories are applied on a per package basis, and since there may be more than one package per source repository, it's recommended that the strictest set of policies apply to the whole repository.
+This is recommended, rather than trying to mix processes depending on which packages are changed in a given change request (pull request or merge request, etc.).
+If this is too onerous, then it's recommended to split lower quality packages out into a separate repository.
+
+
+Motivation
+==========
+
+TBD
+
+
+Rationale
+=========
+
+TBD
+
+
+Backwards Compatibility
+=======================
+
+Since there are no preceeding standards in this space, at least in the ROS community, of which we are aware, there are no backwards compatibility concerns.
+
+
+Reference Implementation
+========================
+
+TBD: Link to "ROS Core" perscription for how to actually achieve each quality level.
+
+
+References and Footnotes
+========================
+
+.. [1] Initial discussions about this REP:
+   https://github.com/ros2/ros2_documentation/pull/460
+
+
+Copyright
+=========
+
+Copyright 2019 Open Source Robotics Foundation, Inc.
```

Comment:
Is this on purpose? All (I believe) other REPs:

> ha[ve] been placed in the public domain.

([REP 2000](https://github.com/ros-infrastructure/rep/blame/190c7a8ecf83293741a0573b59616c5c96f9779c/rep-2000.rst#L584) fi)

Reply:
Yeah, I'm not sure. I placed the license on this as CCA4 because that's what it was from index.ros.org, where I moved it from.

I thought the copyright should be set, even if it should be freely available for reuse.

I don't feel strongly about this, public domain is fine by me actually.

I'll just @gerkey on this to see what he thinks.

Reply:
I think that we should continue our pattern of following the PEP example (which has worked well so far). From [PEP 1](https://www.python.org/dev/peps/pep-0001/):

> Copyright/license -- Each new PEP must be placed under a dual license of public domain and CC0-1.0-Universal.

In that case, we'd use text like this in a REP:

> This document is placed in the public domain or under the CC0-1.0-Universal license, whichever is more permissive.

Reply:
Just to be clear, I'll remove the copyright and license subsections in this document and instead have a "Copyright" subsection (or should it be "Copyright and License"?) which has the PEP recommended text?

Reply:
I propose following PEP's [example](https://www.python.org/dev/peps/pep-0001/#copyright) of including a "Copyright" section whose content is that one sentence.

From browsing PEPs, it seems that they somewhat recently adopted this policy and are not retroactively applying it older PEPs (some older ones have no copyright section, some say just public domain). That also seems like a sensible approach for us.

Reply:
See c2ba03b for the fix. Thanks @gavanderhoorn and @gerkey.